### ISSUE TITLE
Can't send more than 135Kb of data between arm64 hosts(fastdds v2.3.3). [13719]

### ISSUE THREAD

**Issue by large-cat** [eProsima/Fast-DDS#2420]
## Expected Behavior

We want to send 500K data between arm64 hosts.
## Current Behavior

We can send up to 135kb of data between arm64 hosts at one time, but in fastdds v1.5 can send 500kb. 
## Steps to Reproduce

We have configured the QoS to MemoryManagementPolicy::DYNAMIC_RESERVE_MEMORY_MODE and ASYNCHRONOUS_PUBLISH_MODE. 

## System information
arm64 linux and inter host.

**Comment by large-cat** [comment]
Hi @MiguelCompany
Through testing and comparison, we found that the problem occurred in a commits from V1.9 to v1.10. It's `Avoid temporary allocations when not using flow controllers (#1039)[26fe712f6f07672032b12ababeb2e41c137d4887]`. 

However, due to a bug before this commit, there are some exceptions in communication with fastdds, so we need to apply a commit (`Reduce builtin endpoints memory consumption (#1041)[528bd744f523bcc2c29b6e2721174978e259b61b]`) after it to find the real problem.After that, we found the code change that caused the problem. This is because the commit of `26fe712f6f07672032b12ababeb2e41c137d4887` modifies the code of `StatefulWriter.cpp`. 

If we undo this change, we can send 500kb of data, otherwise we can only send data no more than 135kb (arm64 inter host communication). 
So, could you help us fix this problem in fastdds V2.3.3 (because there are a lot of code changes, and we don't know what risks this repair will bring).
Thank you.

**Comment by large-cat** [comment]
Hi @MiguelCompany

Through further testing, we find that the problem is caused by the high packet loss rate (about 10%). After packet loss, the packet loss is aggravated due to the fast retransmission speed of DDS, resulting in congestion in the whole network. Then we make static flow control for the UDP interface of FastDDS v2.3, and then we can send 1MB data between two hosts (the flow control provided by DDS does not work on this issue). So can you give us some suggestions to solve this problem without using static flow control.

On the other hand, we found a bug in the software: if the subscriber history is full of incomplete changes, the flag bit m_isHistoryfull will set to true. However, after a period of time, these changes will be cleared by some mechanism of the software, but the flag bit is m_isHistoryfull is not restored. So that the subscriber can no longer receive messages.
in `SubscriberHistory.cpp`
```cpp
bool SubscriberHistory::add_received_change(
        CacheChange_t* a_change)
{
    if (m_isHistoryFull)
    {
        // Discarding the sample.
        logWarning(SUBSCRIBER, "Attempting to add Data to Full ReaderHistory: " << topic_att_.getTopicDataType());
        return false;
    }
    if (add_change(a_change))
    {
        if (m_changes.size() == static_cast<size_t>(m_att.maximumReservedCaches))
        {
            m_isHistoryFull = true;
        }
        logInfo(SUBSCRIBER, topic_att_.getTopicDataType()
                << ": Change " << a_change->sequenceNumber << " added from: "
                << a_change->writerGUID; );
        return true;
    }
    return false;
}
```
Then we tried to repair it and it can work:
```cpp
bool SubscriberHistory::add_received_change(
        CacheChange_t* a_change)
{
    if (m_isHistoryFull)
    {
        // Discarding the sample.
        if(m_changes.size() < static_cast<size_t>(m_att.maximumReservedCaches)){
            m_isHistoryFull = false;

        } else {
            logWarning(SUBSCRIBER, "Attempting to add Data to Full ReaderHistory: " << topic_att_.topicName);
            return false;
        }
    }

    if (add_change(a_change))
    {
        if (m_changes.size() == static_cast<size_t>(m_att.maximumReservedCaches))
        {
            m_isHistoryFull = true;
        }

        logInfo(SUBSCRIBER, topic_att_.getTopicDataType()
                << ": Change " << a_change->sequenceNumber << " added from: "
                << a_change->writerGUID; );

        return true;
    }

    return false;
}
```

**Comment by MiguelCompany** [comment]
@large-cat Regarding the bug you mention, it was already detected and fixed on #2239.

The backport to 2.3.x is on #2296, which I just merged and will become part of v2.3.5.

Could you explain further what you mean with `static flow control` ?
It seems you tried to use the throughput controller on Fast DDS. Could you explain how you configured it, so we can check if your configuration is correct?

**Comment by large-cat** [comment]
@MiguelCompany
We set flow control Qos by this code:
```cpp
  dataWriterQos->throughput_controller().bytesPerPeriod = 64000*2;
  dataWriterQos->throughput_controller().periodMillisecs = 10;
``` 

And the static `flow contrl` means that : Set an environment variable to describe the bandwidth (100Mbs), and then read this value in the program so that the sending rate of UDP's send function does not exceed this value. 
code:
```cpp
bool UDPTransportInterface::send(
        const octet* send_buffer,
        uint32_t send_buffer_size,
        eProsimaUDPSocket& socket,
        fastrtps::rtps::LocatorsIterator* destination_locators_begin,
        fastrtps::rtps::LocatorsIterator* destination_locators_end,
        bool only_multicast_purpose,
        const std::chrono::steady_clock::time_point& max_blocking_time_point)
{
    static size_t last_send_buff_size = 0;
    static std::chrono::_V2::steady_clock::time_point last_send_time;
    static const char* bandwidth_limitation = getenv("RTPS_BANDWIDTH_LIMITATION");

    fastrtps::rtps::LocatorsIterator& it = *destination_locators_begin;

    bool ret = true;

    auto time_out = std::chrono::duration_cast<std::chrono::microseconds>(
        max_blocking_time_point - std::chrono::steady_clock::now());

    while (it != *destination_locators_end)
    {   
        if(bandwidth_limitation != nullptr && last_send_buff_size != 0){
            static int bandwidth = atoi(bandwidth_limitation);
            if(0 < bandwidth){
                auto time_diff = std::chrono::duration_cast<std::chrono::microseconds>(
                    std::chrono::steady_clock::now() - last_send_time);
            
                //= byte_size x 8 / speed(bandwidth * 1024^2 bit/s)  * 1000^2 
                // = byte_size * (8 * 1000^2 / 1024^2) / bandwidth microseconds
                auto time_interval = std::chrono::microseconds((int)(last_send_buff_size * 7.62939453125 / bandwidth));
                if(time_interval > time_diff){
                    std::this_thread::sleep_for(time_interval - time_diff);
                }
            }
        }
        if (IsLocatorSupported(*it))
        {
            last_send_time = std::chrono::steady_clock::now();
            ret &= send(send_buffer,
                            send_buffer_size,
                            socket,
                            *it,
                            only_multicast_purpose,
                            time_out);
                            
            if(ret) last_send_buff_size = send_buffer_size;
        }

        ++it;
    }

    return ret;
}
```

**Comment by MiguelCompany** [comment]
> We set flow control Qos by this code:
> 
> ```c++
>   dataWriterQos->throughput_controller().bytesPerPeriod = 64000*2;
>   dataWriterQos->throughput_controller().periodMillisecs = 10;
> ```

The throughput controller works better with higher periods because the spent bandwidth is recovered after the period elapses. I would recommend multiplying both values by a certain factor, and perhaps using [XML configuration](https://fast-dds.docs.eprosima.com/en/v2.3.3/fastdds/xml_configuration/common.html#throughput) so you can update those values without rebuilding the application.

**Comment by large-cat** [comment]
> higher periods because the spent bandwidth is recovered after the period elapses. I would recommend multiplying both values by a certain factor, and perhaps using [XML configuration](https://fast-dds.docs.eprosima.com/en/v2.3.3/fastdds/xml_configuration/common.html#throughput) so you can update those values without rebuilding the application.

Hi @MiguelCompany 
Does this problem(#2296 ) exist in 2.6？

**Comment by large-cat** [comment]
> @large-cat Regarding the bug you mention, it was already detected and fixed on #2239.
> 
> The backport to 2.3.x is on #2296, which I just merged and will become part of v2.3.5.
> 
> Could you explain further what you mean with `static flow control` ? It seems you tried to use the throughput controller on Fast DDS. Could you explain how you configured it, so we can check if your configuration is correct?

Hi @MiguelCompany 
Does this problem(#2296 ) exist in 2.6？

**Comment by MiguelCompany** [comment]
Hi @large-cat 

> Does this problem(#2296 ) exist in 2.6？

No. This was resolved starting on 2.4.1. The backport is already merged in 2.3.x

**Comment by large-cat** [comment]
> Hi @large-cat
> 
> > Does this problem(#2296 ) exist in 2.6？
> 
> No. This was resolved starting on 2.4.1. The backport is already merged in 2.3.x

Okay, thank you.

**Comment by Mario-DL** [comment]
I am closing this issue as it seems to be solved. Please, feel free to reopen it if necessary.
